test: Integration tests for v1

stanleyjones opened the pull request:
I'm actually working on integration tests for v2 (i.e. the `next` branch) but they keep failing and I wanted to make sure that the same tests work in v1. So here they are. Honestly, just the basics.

- Can we send an already encoded request and get a response?
- Can we encode a request from bytes and get a response?
- Can we form a request from an object and get a response?
- Can we call a method to form a request and get a response?
- Can we use a shortcut method to form a request and get a response?
- Can we sign a request and get a response (the rest are all anonymous)?

### Test Results

```
❯ npm run test:integration

> @liftedinit/many-js@0.1.0-alpha.9 test:integration
> jest --testMatch '**/?(*.)+(integration).[jt]s?(x)'

 PASS  src/__tests__/network.integration.ts
  Network
    ✓ should send encoded bytes (21 ms)
    ✓ should send a request from bytes (9 ms)
    ✓ should send a request from an object (6 ms)
    ✓ should call a method (6 ms)
    ✓ should call a method from a shortcut (6 ms)
    ✓ should call a method non-anonymously (31 ms)

Test Suites: 1 passed, 1 total
Tests:       6 passed, 6 total
Snapshots:   0 total
Time:        1.189 s
Ran all test suites.
```

### Coverage

From unit tests, not integration.

```
----------------------------------------|---------|----------|---------|---------|------------------------------------
File                                    | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
----------------------------------------|---------|----------|---------|---------|------------------------------------
All files                               |   84.77 |     66.6 |   77.25 |   84.92 |
 src                                    |   57.14 |      100 |     100 |   57.14 |
  const.ts                              |     100 |      100 |     100 |     100 |
  index.ts                              |       0 |      100 |     100 |       0 | 1-3
 src/identity                           |   89.13 |    66.66 |   81.81 |   89.13 |
  address.ts                            |   86.48 |    66.66 |   77.77 |   86.48 | 14,34,65-68
  index.ts                              |     100 |      100 |     100 |     100 |
  types.ts                              |     100 |      100 |     100 |     100 |
 src/identity/anonymous                 |   72.72 |      100 |      40 |      70 |
  anonymous-identity.ts                 |   66.66 |      100 |      25 |   66.66 | 12-20
  index.ts                              |     100 |      100 |     100 |     100 |
 src/identity/ed25519                   |   94.59 |      100 |      80 |   94.59 |
  ed25519-key-pair-identity.ts          |   94.44 |      100 |      80 |   94.44 | 61,79
  index.ts                              |     100 |      100 |     100 |     100 |
 src/identity/webauthn                  |      50 |        0 |   31.25 |   50.84 |
  index.ts                              |     100 |      100 |     100 |     100 |
  webauthn-identity.ts                  |   49.15 |        0 |   31.25 |      50 | 24,33-77,111-184
 src/message                            |   75.75 |    39.24 |   85.71 |   75.75 |
  cbor.ts                               |     100 |      100 |     100 |     100 |
  cose.ts                               |   83.33 |    37.03 |   85.71 |   83.33 | 105-114,127,153
  error.ts                              |   91.66 |    64.28 |     100 |   91.66 | 15
  index.ts                              |   60.78 |    31.57 |   81.81 |   60.78 | 43,48,51,55,62,65,68,77,90-101,114
 src/message/attributes                 |     100 |      100 |     100 |     100 |
  attributes.ts                         |     100 |      100 |     100 |     100 |
  index.ts                              |     100 |      100 |     100 |     100 |
  types.ts                              |     100 |      100 |     100 |     100 |
 src/message/attributes/async-attribute |     100 |       80 |     100 |     100 |
  async-attribute.ts                    |     100 |       80 |     100 |     100 | 15,27
  index.ts                              |     100 |      100 |     100 |     100 |
 src/network                            |    91.3 |    55.55 |     100 |    91.3 |
  index.ts                              |     100 |      100 |     100 |     100 |
  network.ts                            |   90.47 |    55.55 |     100 |   90.47 | 31,35
 src/network/modules                    |     100 |      100 |     100 |     100 |
  index.ts                              |     100 |      100 |     100 |     100 |
  types.ts                              |     100 |      100 |     100 |     100 |
 src/network/modules/account            |   94.87 |    77.77 |   94.73 |   97.33 |
  account.ts                            |   94.73 |    77.77 |     100 |   97.29 | 240,272
  index.ts                              |     100 |      100 |       0 |     100 |
 src/network/modules/async              |   89.58 |    71.42 |     100 |   89.13 |
  async.ts                              |   89.36 |    71.42 |     100 |   88.88 | 51,54,61,91,94
  index.ts                              |     100 |      100 |     100 |     100 |
 src/network/modules/base               |     100 |       75 |   83.33 |     100 |
  base.ts                               |     100 |       75 |     100 |     100 | 55-60
  index.ts                              |     100 |      100 |       0 |     100 |
 src/network/modules/blockchain         |      50 |      100 |       0 |      40 |
  blockchain.ts                         |      25 |      100 |       0 |      25 | 5-15
  index.ts                              |     100 |      100 |       0 |     100 |
 src/network/modules/events             |    92.3 |    73.52 |   88.88 |    92.1 |
  events.ts                             |   91.89 |    73.52 |     100 |   91.89 | 203,218,224
  index.ts                              |     100 |      100 |       0 |     100 |
 src/network/modules/id-store           |     100 |    68.75 |     100 |     100 |
  id-store.ts                           |     100 |    68.75 |     100 |     100 | 33-63,80
  index.ts                              |     100 |      100 |     100 |     100 |
 src/network/modules/kvStore            |   22.22 |        0 |   14.28 |   20.45 |
  index.ts                              |     100 |      100 |       0 |     100 |
  kvStore.ts                            |    18.6 |        0 |   15.38 |    18.6 | 23-87,101-114
 src/network/modules/ledger             |   88.88 |    76.92 |    62.5 |   91.17 |
  index.ts                              |     100 |      100 |     100 |     100 |
  ledger.ts                             |   88.23 |    76.92 |   57.14 |    90.9 | 35-46
 src/network/modules/test               |     100 |       75 |     100 |     100 |
  test-utils.ts                         |     100 |       75 |     100 |     100 | 30
 src/network/modules/tokens             |   96.82 |    68.96 |    87.5 |   96.77 |
  index.ts                              |     100 |      100 |       0 |     100 |
  tokens.ts                             |   96.72 |    68.96 |   93.33 |   96.72 | 66-67
 src/utils                              |    92.3 |    78.26 |    87.5 |    92.3 |
  index.ts                              |    92.3 |    78.26 |    87.5 |    92.3 | 43,52,98
 src/utils/transactions                 |   94.68 |       80 |   94.73 |   94.62 |
  index.ts                              |   94.68 |       80 |   94.73 |   94.62 | 95,210,225-229,276
----------------------------------------|---------|----------|---------|---------|------------------------------------
```

stanleyjones commented on package.json:
```diff
@@ -29,6 +29,7 @@
   },
   "scripts": {
     "test": "jest",
+    "test:integration": "jest --testMatch '**/?(*.)+(integration).[jt]s?(x)'",
```
Integration tests are named `[filename].integration.ts`.

stanleyjones commented on package.json:
```diff
@@ -50,6 +51,10 @@
   "jest": {
+    "collectCoverageFrom": [
+      "./src/**",
+      "!./src/**/__tests__/**"
```
Ignore tests themselves when calculating code coverage.

stanleyjones commented on src/__tests__/network.integration.ts:
```diff
@@ -0,0 +1,54 @@
+import { ID1, SERVERS } from "./data"
+
+describe("Network", () => {
+  const HEX = "d28440a053d92711a2036673746174757305c11a64234ee740"
```
This request is well-formed but will return a Many error because it's out of the accepted time bounds. It's not useful to test the content of the response, just that we received one.

stanleyjones commented on src/message/cose.ts:
```diff
@@ -1,5 +1,5 @@
-import cbor from "cbor";
-import { sha3_224 } from "js-sha3";
+import cbor from "cbor"
+import { sha3_224 } from "js-sha3"
```
Prettier strikes again!

stanleyjones commented on src/message/cose.ts:
```diff
@@ -39,7 +39,9 @@ export class CoseMessage {
   static fromCborData(data: CborData): CoseMessage {
     const cose = cbor.decodeFirstSync(data, decoders).value
-    const protectedHeader = cbor.decodeFirstSync(cose[0])
+    const protectedHeader = cose[0]?.length
+      ? cbor.decodeFirstSync(cose[0])
+      : cose[0]
```
The encoded request we use in tests is anonymous so it has an empty protected header and will throw an error when attempting to decode it. We typically only create messages from CBOR during the parsing of a _response_ (which are always signed) and not a request so we never ran into this situation before.

stanleyjones reviewed (commented):
Some helpful comments...

fmorency reviewed (changes requested):
Thanks, @stanleyjones!

The tests pass when used against the `run.sh` script but fail if used against a `many-ledger` instance (no tendermint/abci) or the docker environment (exposed port points to abci).

I don't think using the `run.sh` and bypassing `many-abci` is correct. 
If using the `run.sh` script, changing
```
LEDGER: "http://127.0.0.1:8001",
```

to

```
LEDGER: "http://127.0.0.1:8000",
```
will hit the abci instead of the ledger. You should be able to reproduce the following error
```
➜  many-js git:(v1-integration-tests) ✗ npm run test:integration 

> @liftedinit/many-js@0.1.0-alpha.9 test:integration
> jest --testMatch '**/?(*.)+(integration).[jt]s?(x)'

ts-jest[versions] (WARN) Version 29.5.0 of jest installed has not been tested with ts-jest. If you're experiencing issues, consider using a supported version (>=28.0.0 <29.0.0-0). Please do not report issues in ts-jest if you are using unsupported versions.
 FAIL  src/__tests__/network.integration.ts
  Network
    ✓ should send encoded bytes (48 ms)
    ✕ should send a request from bytes (14 ms)
    ✓ should send a request from an object (7 ms)
    ✓ should call a method (7 ms)
    ✓ should call a method from a shortcut (7 ms)
    ✓ should call a method non-anonymously (54 ms)

  ● Network › should send a request from bytes

    The message's timestamp is out of the accepted range of the server.

      75 |     const data = content.get(4);
      76 |     if (typeof data == "object" && !Buffer.isBuffer(data)) {
    > 77 |       throw new ManyError(
         |             ^
      78 |         Object.fromEntries(data.entries()) as SerializedManyError
      79 |       );
      80 |     }

      at Function.fromCoseMessage (src/message/index.ts:77:13)
      at Function.fromCborData (src/message/index.ts:86:20)
      at Network.send (src/network/network.ts:40:15)
      at Object.<anonymous> (src/__tests__/network.integration.ts:20:20)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 5 passed, 6 total
Snapshots:   0 total
Time:        1.892 s, estimated 2 s
Ran all test suites.
```

stanleyjones commented:
> Can you craft a properly signed request using a random identity?

Added.

> You should be able to reproduce the following error

Yes, changing the port allowed me to reproduce the error. I've added a means to not get that error (running `many message status --hex` before running the tests and then using the result) but it makes some assumptions about relative paths. We could beef it up if you think it's the right direction.

stanleyjones commented on src/__tests__/network.integration.ts:
```diff
@@ -0,0 +1,69 @@
+  let CBOR_DATA: CborData
+
+  beforeAll(() => {
+    const hex = execSync("../many-rs/target/debug/many message status --hex")
+    HEX = `d2${hex.toString()}`
+    CBOR_DATA = Buffer.from(HEX, "hex")
```
This assumes `many-js` is adjacent to `many-rs` in your file structure, which might not be true.

fmorency replied:
You don't need to rely on `many-rs` for this.

The request will look like this

8440a053d92711a2036673746174757305c1**1a642b22b6**40

where the bytes in **bold** represent the timestamp. This is the only part of that request that changes.

All you need to do is update those bytes with the current cbor-encoded epoch timestamp.

fmorency reviewed (changes requested):
Thanks!
